api/proto/teleport/access_graph/v1/secrets_service.proto

```diff
@@ -0,0 +1,104 @@
+// Copyright 2024 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.access_graph.v1;
+
+import "teleport/access_graph/v1/secrets.proto";
+import "teleport/devicetrust/v1/device_collected_data.proto";
+import "teleport/devicetrust/v1/devicetrust_service.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/accessgraph/v1;accessgraphv1";
+
+// SecretsScannerService provides methods for Access Graph Secret Scanner
+// functionality.
+service SecretsScannerService {
+  // ReportAuthorizedKeys is used by Teleport SSH nodes to report authorized keys
+  // that could be used to bypass Teleport.
+  rpc ReportAuthorizedKeys(stream ReportAuthorizedKeysRequest) returns (stream ReportAuthorizedKeysResponse) {}
+  // ReportSecrets is used by trusted devices to report secrets found on the host
+  // that could be used to bypass Teleport.
+  rpc ReportSecrets(stream ReportSecretsRequest) returns (stream ReportSecretsResponse) {}
+}
+
+// ReportAuthorizedKeys is used by Teleport nodes to report authorized keys
+// that could be used to bypass Teleport.
+message ReportAuthorizedKeysRequest {
+  // keys is a list of authorized keys that could be used to bypass Teleport.
+  repeated teleport.access_graph.v1.AuthorizedKey keys = 1;
+}
+
+// ReportAuthorizedKeysResponse is the response from ReportAuthorizedKeys
+// RPC method.
+message ReportAuthorizedKeysResponse {}
+
+// Request for ReportSecrets.
+//
+// Authentication ceremony flow:
```

Comment:
Why run the authentication ceremony again? This adds a lot of complexity, as we now have to deal with the same ceremony in 2 separate places.

Could we have the client authenticate by regular means and check the connecting certificate for device extensions instead?

Reply:
We do not repeat the authentication process here. The objective is to use the device itself as credentials to access Teleport. This approach offers several significant benefits:

- It eliminates the need for valid user logins during the scan procedure. The device acting as credentials addresses these challenges because:
  1. Scans can be performed outside of working hours.
  2. The device can be correctly and uniquely identified when the MDM solution runs the scan in the inventory.

-   The device has an owner and tracks users who have logged in from it and all that information is tracked in Teleport already.

- We ensure that the connecting device is trusted, preventing connections from being relayed through unauthorized third-party laptops.

- It ensures continuity of operations, as devices can be authenticated and authorized even in scenarios where user credentials might be compromised, locked or unavailable. 


Devices will not connect to Auth using a certificate-key pair. Instead, they will establish a connection to Auth without credentials and gain access to the endpoint by solving a challenge to prove their identity. Auth already supports credentialless connections.

Reply:
> We do not repeat the authentication process here

We do. The way to make this work is to make parts of e/lib/devicetrust/devicetrustv1 public (which they were never meant to be) and "borrow" logic from it, effectively forcing SecretsScannerService to require the same accesses (and "powers") as the DeviceTrustService itself.

Service-to-service communication should happen through the gRPC layer, so we keep service isolation and responsibilities intact. This pays off long-term if we want to split service deployment, isolate sensitive database keys, and even in managing complexity (as device trust logic isn't suddenly under SecretsScannerService).

I have suggested a few alternatives, which I'll restate here:

1. Add a DeviceTrustService.AssertDevice RPC. We get a token from it that proves the device was verified. Yes, the token could be moved, but it still seems acceptable to me for what SecretsScannerService does (and other functionality could benefit).

2. Do a lightweight key-possession challenge on SecretsScannerService, without borrowing logic from DeviceTrustService, then use DeviceTrustService.GetDevice or FindDevices to verify the asset_tag and public_key. This doesn't break service isolation and I'd be happy to provide APIs under lib/devicetrust to do that.

As this stands I wouldn't be happy approving the changes myself.


> This approach offers several significant benefits:

This seems fine, I've no problems with the bullets listed there. I think we can achieve all of them while keeping service isolation intact.

>> We ensure that the connecting device is trusted, preventing connections from being relayed through unauthorized third-party laptops.

Note that if someone hacked a trusted device to a point where it can be driven from afar, they could do whatever they wanted with TPM-based challenges. macOS stands out here because of the need for signed binaries.

Reply:
Discussed out of band with Tiago, we've reached a good solution after combining various ideas (from here and otherwise).

I'll resolve the thread. Thanks for the interesting discussions, Tiago!